<a href="https://colab.research.google.com/github/shravanineeli/ML_Projects/blob/main/Transition_States.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Transition Matrix

Matrix -

    .         Present   Late   Absent
    Present    20/27    5/27    2/27
    Late       8/21     10/21   3/21
    Absent     5/21     1/21    15/21

In [ ]:
#importing libraries
import numpy as np

#define transition matrix
P = np.array([[20/27, 5/27, 2/27], [8/21, 10/21, 3/21], [5/21, 1/21, 15/21]])
outcomes = ["Present", "Late", "Absent"]

#checking the matrix
print(P)

[[0.74074074 0.18518519 0.07407407]
 [0.38095238 0.47619048 0.14285714]
 [0.23809524 0.04761905 0.71428571]]


# Probabilities

In [ ]:
#initial state vector
initial_state = np.array([1, 0, 0]) #Present

#Probabilities of trasitioning from present
state_distribution_present = np.dot(initial_state,
                                            np.linalg.matrix_power(P, 2))
state_distribution_present

array([0.63688027, 0.22888497, 0.13423476])

The probability of a student being present and then present is 63.69%, present and then late is 22.89%, and lastly, present and then absent is 13.42%.

In [ ]:
#initial state vector
initial_state = np.array([0, 1, 0]) #Late

#Probabilities of transitioning from late
state_distribution_late = np.dot(initial_state,
                                            np.linalg.matrix_power(P, 2))
state_distribution_late

array([0.49760645, 0.30410683, 0.19828672])

The probability of a student being late and then present is 49.76%, late and then late is 30.41%, and lastly, late and then absent is 19.83%.

In [ ]:
#initial state vector
initial_state = np.array([0, 0, 1]) #Absent

#Probabilities of transitioning from absent
state_distribution_absent = np.dot(initial_state,
                                            np.linalg.matrix_power(P, 2))
state_distribution_absent

array([0.36457546, 0.10078105, 0.53464349])

The probability of a student being absent and then present is 36.46%, absent and then late is 10.08%, and lastly, absent and then absent is 53.46%.

# Average Number of Days to Transition from Present to Late

In [ ]:
#initial state vector
initial_state = np.array([1, 0, 0]) #Present

#displaying the initial state
print("Initial State:", outcomes[np.argmax(initial_state)])

Initial State: Present


In [ ]:
#calculating average number of days for a student to transition from Present to Late
day = 1
max_days = 27
while np.argmax(initial_state) != 1 and day <= max_days:
  initial_state = np.dot(initial_state, P)
  print(f"Day {day}: {outcomes[np.argmax(initial_state)]}")
  day += 1

Day 1: Present
Day 2: Present
Day 3: Present
Day 4: Present
Day 5: Present
Day 6: Present
Day 7: Present
Day 8: Present
Day 9: Present
Day 10: Present
Day 11: Present
Day 12: Present
Day 13: Present
Day 14: Present
Day 15: Present
Day 16: Present
Day 17: Present
Day 18: Present
Day 19: Present
Day 20: Present
Day 21: Present
Day 22: Present
Day 23: Present
Day 24: Present
Day 25: Present
Day 26: Present
Day 27: Present


If a student is present on Day 1, they are never late.

# Average Number of Days to Transition from Late to Absent

In [ ]:
#initial state vector
initial_state = np.array([0, 1, 0]) #Late

#displaying the initial state
print("Initial State:", outcomes[np.argmax(initial_state)])

Initial State: Late


In [ ]:
#calculating average number of days for a student to transition from Late to Absent
day = 1
max_days = 21
while np.argmax(initial_state) != 2 and day <= max_days:
  initial_state = np.dot(initial_state, P)
  print(f"Day {day}: {outcomes[np.argmax(initial_state)]}")
  day += 1

Day 1: Late
Day 2: Present
Day 3: Present
Day 4: Present
Day 5: Present
Day 6: Present
Day 7: Present
Day 8: Present
Day 9: Present
Day 10: Present
Day 11: Present
Day 12: Present
Day 13: Present
Day 14: Present
Day 15: Present
Day 16: Present
Day 17: Present
Day 18: Present
Day 19: Present
Day 20: Present
Day 21: Present


It takes a student approximately 1 day to transition from Late to Present.

# Average Number of Days to Transition from Absent to Present

In [ ]:
#initial state vector
initial_state = np.array([0, 0, 1]) #Absent

#displaying the initial state
print("Initial State:", outcomes[np.argmax(initial_state)])

Initial State: Absent


In [ ]:
#calculating average number of days for a student to transition from Absent to Present
day = 1
max_days = 21
while np.argmax(initial_state) != 0 and day <= max_days:
  initial_state = np.dot(initial_state, P)
  print(f"Day {day}: {outcomes[np.argmax(initial_state)]}")
  day += 1

Day 1: Absent
Day 2: Absent
Day 3: Present


It takes a student approximately 2 day to transition from Absent to Present.

# Average Number of Days to Transition from Absent to Late

In [ ]:
#initial state vector
initial_state = np.array([0, 0, 1]) #Absent

#displaying the initial state
print("Initial State:", outcomes[np.argmax(initial_state)])

Initial State: Absent


In [ ]:
#calculating average number of days for a student to transition from Absent to Late
day = 1
max_days = 21
while np.argmax(initial_state) != 1 and day <= max_days:
  initial_state = np.dot(initial_state, P)
  print(f"Day {day}: {outcomes[np.argmax(initial_state)]}")
  day += 1

Day 1: Absent
Day 2: Absent
Day 3: Present
Day 4: Present
Day 5: Present
Day 6: Present
Day 7: Present
Day 8: Present
Day 9: Present
Day 10: Present
Day 11: Present
Day 12: Present
Day 13: Present
Day 14: Present
Day 15: Present
Day 16: Present
Day 17: Present
Day 18: Present
Day 19: Present
Day 20: Present
Day 21: Present


If a student is absent on the first day, they are never late.